src/opm/output/eclipse/RestartValue.cpp

```diff
@@ -0,0 +1,96 @@
+/*
+  Copyright (c) 2018 Statoil ASA
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <set>
+#include <iostream>
+
+#include <opm/output/eclipse/RestartValue.hpp>
+
+namespace Opm {
+
+    namespace {
+
+        const std::set<std::string> reserved_keys = {"LOGIHEAD", "INTEHEAD" ,"DOUBHEAD", "IWEL", "XWEL","ICON", "XCON" , "OPM_IWEL" , "OPM_XWEL", "ZWEL"};
+
+    }
+
+
+
+    RestartValue::RestartValue(data::Solution sol, data::Wells wells_arg) :
+        solution(std::move(sol)),
+        wells(std::move(wells_arg))
+    {
+    }
+
+    const std::vector<double>& RestartValue::get_extra(const std::string& key) const {
+        const auto iter = std::find_if(this->extra.begin(), this->extra.end(), [&](std::pair<RestartKey, std::vector<double>> pair) {return (pair.first.key == key);});
+        if (iter == this->extra.end())
+            throw std::invalid_argument("No such extra key " + key);
+
+        return iter->second;
+    }
+
+    bool RestartValue::has_extra(const std::string& key) const {
+        const auto iter = std::find_if(this->extra.begin(), this->extra.end(), [&](std::pair<RestartKey, std::vector<double>> pair) {return (pair.first.key == key);});
+        return  (iter != this->extra.end());
+    }
+
+    void RestartValue::add_extra(const std::string& key, UnitSystem::measure dimension, std::vector<double> data) {
```

Comment:
I assume that it is intentional to take `data` by value, to copy it. That copy should however be moved into `this->extra` instead of copied. One attempt at line 67 (not tested to verify that the copy is dropped):
```
this->extra.emplace_back(RestartKey(key, dimension), std::move(data));
```
I am not 100% sure that gets rid of copying, because of the pair. Might be easier to just use a good old const reference.

Reply:
Yes - it is value by intention; because of the multithreaded output. Don't know 100% if it is required - but it is at least safe.

Reply:
But there is no need to then copy the data again into the pair inside the container?